proj_eclipse/src/main/java/smartmag/models/BoxModel.java

```diff
@@ -0,0 +1,245 @@
+package smartmag.models;
+
+import static ingsw_proj_magazzino.db.generated.Tables.BOX;
+import static ingsw_proj_magazzino.db.generated.Tables.PRODOTTO;
+
+import java.sql.SQLIntegrityConstraintViolationException;
+import java.util.HashMap;
+
+import ingsw_proj_magazzino.db.generated.tables.records.BoxRecord;
+import ingsw_proj_magazzino.db.generated.tables.records.ProdottoRecord;
+import smartmag.data.Box;
+import smartmag.data.Prodotto;
+
+public class BoxModel extends BaseModel {
+
+	private static HashMap<String, BoxModel> instances = new HashMap<String, BoxModel>();
+	private Box box;
+	private BoxRecord record;
+
+	/**
+	 * restituisce il modello del box passato come parametro e lo salva
+	 * all'interno dell'hashmap che contiene i modelli di tutti i box
+	 * 
+	 * @param b box del quale si vuole ottenere il modello
+	 * @return modello del box
+	 */
+	public static BoxModel getBoxModel(Box b) {
+		if (b != null && b.isValid()) {
+			if (!instances.containsKey(b.getIndirizzo())) {
+				if ((DSL.select().from(PRODOTTO)
+						.where(PRODOTTO.ID.eq(b.getProd().getId()))) != null) {
+					BoxModel bm = new BoxModel(b);
+					instances.put(b.getIndirizzo(), bm);
+					return bm;
+				} else {
+					throw new IllegalArgumentException(
+							"prodotto non esistente");
+				}
+
+			} else {
+				return instances.get(b.getIndirizzo());
+			}
+		} else
+			throw new IllegalArgumentException("box non valido");
+	}
+
+	/**
+	 * metodo costruttore della classe BoxModel
+	 * 
+	 * @param b box del quale si vuole generare il modello
+	 */
+	private BoxModel(Box b) {
```

Comment:
Secondo me nel costruttore si dovrebbe inserire nella HashMap instances il modello che si è appena costruito.

Reply:
per come l'ho pensata io lo inserisco nel metodo getBoxModel che viene chiamato per ottenere il modello, lui stesso chiama il costruttore se il modello non esiste già e nell'istruzione successiva inserisce il modello nell'HashMap. dici che cambia qualcosa? secondo me no

Reply:
L'unica differenza è che così devi richiamare il metodo getBoxModel oltre che il costruttore per inserire il nuovo modello nella HashMap. Però il risultato finale è lo stesso.

Reply:
ma il modello lo ottieni solo tramite getbBoxModel no?

Reply:
perché il costruttore é privato

Reply:
Si si, potresti fare che con getBoxModel te lo ritorna il record mentre nel costruttore te lo inserisce solo nella HashMap

Reply:
Sì, in effetti anche io ho centrallizzato la modifica della mappa nel costruttore come ha detto Davide. Questo perché se dovessi modificare il modello rischi meno errori e non devi pensarci. Guarda pure anche i nostri modelli come riferimento